### PR TITLE
Api fixes

### PR DESCRIPTION
**Pull request by MeirShpilraien**
Introduce the following api changes:

1. KeysOnlyReader is implemented as python reader and support the following arguments:
```
	def CreateKeysOnlyReader(pattern='*', count=1000, exactMatch=False, patternGenerator=None):
    	'''
    	Create a KeysOnlyReader callback as a python reader
    	pattern          - keys pattern to return
    	count            - count parameter given to scan command (ignored if isPattern is false)
    	exactMatch       - boolean indicating if exact match is require, i.e, do not use scan and just return the
                       give pattern
    	patternGenerator - a callbacks to generate different pattern on each shard. If given, the callback
                       will run on each shard and the return tuple (pattern, isPattern) will be used.
                       If this argument is given the pattern and isPattern arguments are ignored.
    	'''
```
2. KeysReader format was changed to (notice there is no backword issues here): {'key':..., 'value':..., 'type':..., 'event':...}
   * key - the key name (string)
   * value - the key value (type is depends on the key type)
   * type - the key type (string, hash, set, ...)
   * event - the event that trigger the exection (SET, HSET, ..), None on batch executions

   In addition the KeysReader introduce the following new/changed capabilities:
   * on 'run':
      * readValue - if False only the 'key' and the 'event' values will be provided (default True)
      * noScan - if True read the exect pattern given without performing scan operation (default False)
   * on 'register':
      * readValue - same as read value on 'run'
      * regex argument was depricated and changed to prefix

3. StreamReader format was changed to (notice there is a backword compatability): {'key':..., 'msg_id':..., 'value':{'k':'v',...}}
   * key - the stream key (string)
   * msg_id - the current message id (string)
   * value - the message value (dictionary)

   In addition the StreamReader introduce the following new/changed capabilities:
   * on 'run':
      * fronId - the id from which to start read messages from the stream (default 0-0)
   * on 'register':
      * regex argument was depricated and changed to prefix

4. the 'Log' function was changed to 'Log(msg, level='notice')'.
   If old argument format is used a depricated message is added to the log file

5. Pytest was added to cover all the new arguments

Still left to do:
- [x] Add tests for KeysOnlyReader
- [x] Document all the changes
- [x] Delete the KeysOnlyReader from Keys_reader.c
- [x] Change msg_id to just id

@itamarhaber FYI

### CODE REVIEW

File: src/GearsBuilder.py [hunk 1]
```diff
@@ -15,6 +15,39 @@
...
+def CreateKeysOnlyReader(pattern='*', count=1000, exactMatch=False, patternGenerator=None):
+    '''
+    Create a KeysOnlyReader callback as a python reader
+    pattern          - keys pattern to return
+    count            - count parameter given to scan command (ignored if isPattern is false)
+    exactMatch       - boolean indicating if exact match is require, i.e, do not use scan and just return the
+                       give pattern
+    patternGenerator - a callbacks to generate different pattern on each shard. If given, the callback
+                       will run on each shard and the return tuple (pattern, isPattern) will be used.
+                       If this argument is given the pattern and isPattern arguments are ignored.
```
Comment (rafie) [inline src/GearsBuilder.py]: isPattern => exactMatch?

File: src/readers/keys_reader.c [hunk 2]
```diff
@@ -124,78 +125,81 @@ static KeysReaderRegisterData* KeysReaderRegisterData_GetShallowCopy(KeysReaderR
...
+    RedisGears_BWWriteLong(bw, krctx->noScan);
+}
+
+static void RG_KeysReaderCtxDeserialize(FlatExecutionPlan* fep, void* ctx, Gears_BufferReader* br){
+    KeysReaderCtx* krctx = (KeysReaderCtx*)ctx;
+    krctx->match = RG_STRDUP(RedisGears_BRReadString(br));
+    if(RedisGears_BRReadLong(br)){
+        krctx->event = RG_STRDUP(RedisGears_BRReadString(br));
+        // having an even means this execution was triggered by a registration and the fact
+        // that we deserialize it means that the event did not accured on this shard,
```
Comment (rafie) [inline src/readers/keys_reader.c]: accured => occurred

**Comment by MeirShpilraien** [comment]
@rafie @itamarhaber can this be merged?

**Comment by itamarhaber** [comment]
I'd like to also suggest:
[ ] Decapitalize `Log()` to `log()`
[ ] Ditto `ConfigGet` to `configGet()`

File: docs/functions.md [hunk 3]
```diff
@@ -205,7 +205,7 @@ The **Register** action registers a function as an event handler. The function i
 ```python
 class GearsBuilder.register(regex='*', mode='async', batch=1, duration=0,
   eventTypes=None, keyTypes=None, onRegistered=None, onFailedPolicy="continue",
-  onFailedRetryInterval=1)
+  onFailedRetryInterval=1, trigger=None)
```
Comment (MeirShpilraien) [inline docs/functions.md]: @itamarhaber I am not sure about this session anymore, basically, the register gets a kargs now and each reader extracts what it needs.

**Comment by MeirShpilraien** [comment]
@itamarhaber done

**Comment by MeirShpilraien** [comment]
@rafie @itamarhaber let me know if I can merge

File: docs/functions.md (hunk 3 above)
Comment (itamarhaber) [reply docs/functions.md]: I see.

So, are there any args that apply to all readers here?

**Comment by itamarhaber** [comment]
lgtm :)

File: docs/functions.md (hunk 3 above)
Comment (MeirShpilraien) [reply docs/functions.md]: Only the `onRegistered=None`, `mode`, `regex (which was changed to prefix)`
I will fix it

Comment (MeirShpilraien) [reply docs/functions.md]: Know what, its actually just the `onRegistered=None` because of command reader

**Pull request merged**